File: 967-minimum-falling-path-sum/minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        for (int i = 1; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] += Math.min(matrix[i-1][Math.max(j-1, 0)], 
                                         Math.min(matrix[i-1][j], 
                                                  matrix[i-1][Math.min(j+1, cols-1)]));
            }
        }

        int minSum = Integer.MAX_VALUE;
        for (int num : matrix[rows-1]) {
            minSum = Math.min(minSum, num);
        }

        return minSum;
    }
}

File: 967-minimum-falling-path-sum/minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        // int m = matrix.size(), n = matrix[0].size();
        // if (m == 1) {
        //     return *min_element(matrix[0].begin(), matrix[0].end());
        // }

        // for (int i = 1; i < m; i++) {
        //     for (int j = 0; j < n; j++) {
        //         int min_val = matrix[i - 1][j];
        //         if (j > 0) {
        //             min_val = min(min_val, matrix[i - 1][j - 1]);
        //         }
        //         if (j < n - 1) {
        //             min_val = min(min_val, matrix[i - 1][j + 1]);
        //         }
        //         matrix[i][j] += min_val;
        //     }
        // }

        // return *min_element(matrix[m - 1].begin(), matrix[m - 1].end());
        int m = matrix.size(), n = matrix[0].size();
        if (m == 1)
            return *min_element(matrix[0].begin(), matrix[0].end());
        for (int i = 1; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int minV = matrix[i - 1][j];
                if (j <n-1)
                    minV = min(minV, matrix[i - 1][j + 1]);
                if (j > 0)
                    minV = min(matrix[i - 1][j -1], minV);
                matrix[i][j] += minV;
            }
        }
        return *min_element(matrix[n - 1].begin(), matrix[n - 1].end());
    }
};
